### ISSUE TITLE
Need for better containerization of Benchmark containers

### ISSUE THREAD

**Issue by NeoChaos12** [automl/HPOBench#93]
While trying to run experiments for a recent project on the HPC cluster NEMO, I realized that the containerization strategy of `HPOBench` can use some improvements that would otherwise cause problems for the library's usage in more generic environments than the in-house cluster. The basic issue is that the containers are not yet self-contained enough to fully benefit from the benefits of containerization.

In my particular case, this came to the fore because a new containerized benchmark that I had implemented was unable to read the `.hpobenchrc` file since the process running it did not have the appropriate file-access permissions on the cluster due to the cluster's security policies. This, unfortunately, wasted a lot of time in debugging to identify as not an issue with how the benchmark had been implemented within the framework of `HPOBench` but how the library itself functions. It is also particularly frustrating since the benchmark itself is otherwise fully self-contained and does not rely on any file-reads outside the container. Let me elaborate.

For readability, I will append the exact error logs to the end of this post. The NEMO cluster itself is perfectly fine with running my new benchmark from a job that directly invokes Singularity since it provides the job process and Singularity itself with all the appropriate file-access permissions. However, the cluster places extra restrictions on sub-processes spawned by an already running job process (which makes sense as a number of known exploits use such vulnerabilities). The library I'm using to run my experiments (automl/HPOBenchExperimentUtils) handles arbitrary `HPOBench` benchmarks by invoking a singularity container from an asynchronous sub-process that the main process then interfaces with. This use-case is, in my opinion, perfectly valid, since it enables the library to interface with any arbitrary `HPOBench` benchmark. However, as you may be able to infer from the following error log, it runs into the above mentioned issue and fails.

My personal solution for this is two-fold:
- On the one hand, such configuration files as the `.hpobenchrc` seem to have no real business being read in from outside the container. If this was intended as a way to pass some configuration values to the containerized benchmark, then I would recommend that a better way to do so is to include the configuration values as a part of the container initialization process, e.g. environment variables. Otherwise, the only purpose this serves is to break containerization for benchmarks that do not need to do so.
- Explicitly define how benchmarks should access the outside file-system in case they need it and declare if they even need to. This would take the form of specific directories within the containerized filesystem that would be bind-mounted to the outside file-system if and only if the benchmark requires it. Doing so would go a long way towards maintaining extensibility and re-usability of the library since generic code for handling the library's benchmarks could be written without taking into consideration exactly how each benchmark interacts with the outside world. The current system of data loaders addresses this somewhat, but the protocol is still rather vague. Most notably, it lacks a "switch" to declare if outside file-system access is needed in the first place.

> [INFO] __main__ at 2021-03-17 22:10:48,966 --- Start running benchmark bbobmixint with optimizer setting randomsearch.
> [DEBUG] __main__ at 2021-03-17 22:10:49,178 --- Settings loaded: {'optimizer': 'randomsearch', 'display_name': 'random search', 'num_iterations': 10000000, 'i
> s_surrogate': False, 'tae_limit': None, 'fuel_limit': None, 'time_limit_in_s': 7200, 'cutoff_in_s': 7200, 'mem_limit_in_mb': 6000, 'import_from': 'synthetic.b
> bob_benchmarks', 'import_benchmark': 'BBOBMixIntBenchmark', 'fake_fidelity': True}
> [DEBUG] __main__ at 2021-03-17 22:10:49,178 --- Optimizer: randomsearch
> [DEBUG] __main__ at 2021-03-17 22:10:49,200 --- Output dir: /work/ws/nemo/fr_ab771-bbob-0/experiments/data/9813056/bbobmixint/randomsearch/run-0
> [DEBUG] HPOBenchExperimentUtils.utils.runner_utils at 2021-03-17 22:10:49,201 --- Try to execute command: from hpobench.container.benchmarks.synthetic.bbob_be
> nchmarks import BBOBMixIntBenchmark
> [DEBUG] HPOBenchExperimentUtils.utils.runner_utils at 2021-03-17 22:10:49,262 --- Benchmark BBOBMixIntBenchmark successfully loaded
> [DEBUG] BenchmarkClient at 2021-03-17 22:10:49,263 --- Use benchmark BBOBMixIntBenchmark from container library://phmueller/automl/bbob_benchmarks. 
> And container directory /work/ws/nemo/fr_ab771-bbob-0/.cache/hpobench/hpobench-905753
> [DEBUG] oslo_concurrency.lockutils at 2021-03-17 22:10:49,272 --- Lock "not_thread_process_safe" acquired by "hpobench.container.client_abstract_benchmark.Abs
> tractBenchmarkClient._setup.<locals>.download_container" :: waited 0.009s
> [DEBUG] BenchmarkClient at 2021-03-17 22:10:49,274 --- Skipping downloading the container. It is already downloaded.
> [DEBUG] oslo_concurrency.lockutils at 2021-03-17 22:10:49,275 --- Lock "not_thread_process_safe" released by "hpobench.container.client_abstract_benchmark.Abs
> tractBenchmarkClient._setup.<locals>.download_container" :: held 0.003s
> [DEBUG] BenchmarkClient at 2021-03-17 22:10:49,275 --- SINGULARITYENV_HPOBENCH_DEBUG=true singularity instance start --bind /var/lib/,/work/ws/nemo/fr_ab771-b
> bob-0/.local/share/hpobench:/var/lib/,/work/ws/nemo/fr_ab771-bbob-0/.cache/hpobench/hpobench-905753 /work/ws/nemo/fr_ab771-bbob-0/.cache/hpobench/hpobench-905
> 753/bbob_benchmarks 3f8dc1b0-8765-11eb-88c0-a4bf0104eb90
> [DEBUG] BenchmarkClient at 2021-03-17 22:10:50,156 --- b'\x1b[34mINFO:   \x1b[0m instance started successfully\n'
> [DEBUG] BenchmarkClient at 2021-03-17 22:10:52,206 --- singularity run instance://3f8dc1b0-8765-11eb-88c0-a4bf0104eb90 BBOBMixIntBenchmark 3f8dc1b0-8765-11eb-
> 88c0-a4bf0104eb90
> [DEBUG] BenchmarkClient at 2021-03-17 22:10:52,213 --- Check connection to container and init benchmark
> [DEBUG] BenchmarkClient at 2021-03-17 22:10:52,214 --- Still waiting
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.7/pathlib.py", line 1273, in mkdir
>     self._accessor.mkdir(self, mode)
> FileNotFoundError: [Errno 2] No such file or directory: '/work/ws/nemo/fr_ab771-bbob-0/.config/hpobench'
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/home/HPOBench/hpobench/container/server_abstract_benchmark.py", line 20, in <module>
>     from hpobench.config import HPOBenchConfig
>   File "/usr/local/lib/python3.7/site-packages/hpobench/__init__.py", line 4, in <module>
>     from hpobench.config import config_file  # noqa: F401
>   File "/usr/local/lib/python3.7/site-packages/hpobench/config.py", line 166, in <module>
>     config_file = HPOBenchConfig()
>   File "/usr/local/lib/python3.7/site-packages/hpobench/config.py", line 61, in __init__
>     self._setup()
>   File "/usr/local/lib/python3.7/site-packages/hpobench/config.py", line 75, in _setup
>     self.__create_config_file()
>   File "/usr/local/lib/python3.7/site-packages/hpobench/config.py", line 90, in __create_config_file
>     self.__check_dir(self.config_file.parent)
>   File "/usr/local/lib/python3.7/site-packages/hpobench/config.py", line 163, in __check_dir
>     raise e
>   File "/usr/local/lib/python3.7/site-packages/hpobench/config.py", line 160, in __check_dir
>     Path(path).mkdir(exist_ok=True, parents=True)
>   File "/usr/local/lib/python3.7/pathlib.py", line 1277, in mkdir
>     self.parent.mkdir(parents=True, exist_ok=True)
>   File "/usr/local/lib/python3.7/pathlib.py", line 1273, in mkdir
>     self._accessor.mkdir(self, mode)
> OSError: [Errno 30] Read-only file system: '/work/ws/nemo/fr_ab771-bbob-0/.config'
> [DEBUG] BenchmarkClient at 2021-03-17 22:10:57,220 --- Still waiting
> [DEBUG] BenchmarkClient at 2021-03-17 22:11:02,226 --- Still waiting
> [DEBUG] BenchmarkClient at 2021-03-17 22:11:07,232 --- Still waiting
> [DEBUG] BenchmarkClient at 2021-03-17 22:11:12,239 --- Still waiting

**Comment by PhMueller** [comment]
Hey, thanks for your feedback. 

Since the problem occurs when the container tries to "create" the directory of the configuration file, does this directory already exists on the filesystem? ('/work/ws/nemo/fr_ab771-bbob-0/.config'). Can you please check this? Or maybe create the directory and the config file beforehand? 
It sounds like the problem occurs by using subprocesses with strict cluster policies. In the Hpobench, we spawn the container in a subprocess (https://github.com/automl/HPOBench/blob/4fdd85c9acacbfad73a42a1b32afa44704733074/hpobench/container/client_abstract_benchmark.py#L189-L191). 

Furthermore, I was already thinking about adding an option to bind further directories into the container. In contrast to docker, Singularity melts the file systems together. Thus, if we bind the right paths into the container, we could gain some more flexibility and maybe resolve the file access problems. IMO, binding the config file and inserting environment variables into the container comes to the same result. I think this class of problems can be fixed by binding some more directories without changing the architecture. 
 
 I will discuss it with Katharina on Wednesday.

**Comment by NeoChaos12** [comment]
Hi Philipp,

Checking if the directories existed was the first thing I did. These are the same directories as are being used by the non-containerized `HPOBench` code running on the client side. The `.hpobenchrc` file as well as the directories all existed and were accessible by a container run directly from a job process. As per my analysis of the code and the data flow here, the issue is occurring because the container's sub-process does not even have read access to the filesystem outside the container. Therefore, when it checks for the existence of the `.hpobenchrc` file, it is unable to locate it, and subsequently attempts to create a fresh one. It is unlikely that bind-mounting different directories would fix this issue since the directory `/work/ws/nemo/fr_ab771-bbob-0/` is the only filesystem that my jobs on NEMO are supposed to read/write data from/to.
https://github.com/automl/HPOBench/blob/89ff0392c1b3726d0231d6752996de45ad705487/hpobench/config.py#L73-L75

Again, to clarify, when I say "a container run directly from a job process", I mean a job that was launched with something along the lines of `singularity exec <container>` directly within the job's list of terminal commands to run, as opposed to a more indirect approach towards launching the container by first launching a python process and then launching a sub-process from within it, as occurs in the case of `HPOBenchExperimentUtils`.

It is possible that this mode of usage requires some additional settings for Singularity/NEMO jobs, but I am definitely not aware of what they could be nor could I find anything relevant.

Regarding the config file, there is an additional benefit to what I suggested. A container developer could design a `.hpolibrc` to work exactly how they want it to with some default settings that are intended for the container environment only, whereas someone launching a container could then modify/overwrite a small number of these settings using environment variables. I believe this is, on the whole, much more stable as well, since it makes decoupling the `.hpobenchrc` at the client and server sides easier. Typically, the client should be completely agnostic to the server's `.hpobenchrc`.
